Flight recoder data as JSON

c-p-i-o opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129712
* __->__ #129505

Summary:
Provide a new API to retrieve flight recorder data as JSON.
The one minor difference between flight recorder as Pickle v/s JSON is
that the JSON API does not retrieve stack traces at the moment.
This ends up being far too much data.

Test Plan:
unit test

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Differential Revision: [D59536460](https://our.internmc.facebook.com/intern/diff/D59536460)

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

d4l3k commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3533,13 +3534,25 @@ def test_short(self, timing_enabled, include_collectives):
+                        includeCollectives=False,
+                        includeStackTraces=None,
+                        onlyActive=None,
+                    )
```
This is a bit messy -- maybe better to pull the bulk of this test into a helper method and have separate top level tests for the different types of data?

d4l3k commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -3203,6 +3203,23 @@ such as `dist.all_reduce(tensor, async_op=True)`.
+      "_dump_nccl_trace_json",
+      [](std::optional<bool> includeCollectives,
+         std::optional<bool> onlyActive) {
+        return py::bytes(::c10d::dump_nccl_trace_json(
```
Should we be releasing the GIL during dumping?

c-p-i-o replied:
Unsure about this.

d4l3k commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -924,6 +1004,44 @@ struct NCCLTraceBuffer {
     return pg_config;
   }
 
+  const std::map<std::string, std::map<std::string, std::string>>
```
Should this logic be in a cpp file instead of a header?

c-p-i-o replied:
Moved all of NCCLTraceBuffer struct methods into .cpp

c-p-i-o replied:
These methods are part of `NCCLTraceBuffer` struct. I will move these into the .cpp file in a separate diff to make this diff easier to review.

c-p-i-o committed:
Update

[ghstack-poisoned]

d4l3k commented on torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -365,20 +366,45 @@ std::string dump_nccl_trace(
     std::string ncclUniqueIDStr = buildNcclUniqueIdStr(ncclComm->getNcclId());
     ncclDumpMap[ncclUniqueIDStr] = ncclComm->ncclCommDump();
   }
+  if (json) {
```
This `if (json)` feels like a code smell, can we instead just have two separate methods and move the logic above into a helper that only returns non-null if `IS_NCCLX`?

That would also significantly reduce the amount of code under the `ifdefs` since we don't have two `dump_nccl_trace_helper` impls. It'd be just a `getNCCLDumpMap`

c-p-i-o replied:
Good suggestion! fixed.

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

c-p-i-o committed:
Update

[ghstack-poisoned]

wconstab commented on test/distributed/test_c10d_nccl.py:
```diff
@@ -3592,6 +3567,63 @@ def test_short(self, timing_enabled, include_collectives):
+            include_collectives=include_collectives,
+            timing_enabled=timing_enabled,
+            is_json=True,
+        )
```
nit: we dropped the verification that we had the expected number of entries (2) for this test.  It might be ok to add it back into the helper fn as an argument (expectedLen) or else just add it as a separate check on this test if you don't think it makes sense in the helper

c-p-i-o replied:
We had this:
```
            self.assertEqual(len(t["entries"]), 2)
            t = t["entries"]
            self.assertEqual(len(t), 2). <<<<  THIS CHECK IS DUPLICATE
 ```
 I removed the 3rd line because we were already checking it 2 lines above.

wconstab commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -805,6 +813,77 @@ struct NCCLTraceBuffer {
+      if (onlyActive && e.time_discovered_completed_.has_value()) {
+        continue;
+      }
+      j[record_id_key_str] = int64_t(e.id_);
```
just curious- is it not easy to get an ivalue as a str() via some method?  I guess its still nicer the way you did it.

c-p-i-o replied:
Yes, not easy since `ivalue` is a `generic`. I'd have to write a lot of code in `ivalue` to make it happen.

wconstab commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -923,6 +1001,44 @@ struct NCCLTraceBuffer {
+    }
+
+    if (ncclDumpMap.has_value()) {
+      result[nccl_comm_key_str] = ncclDumpMap.value();
```
is ncclDumpMap.value() another json encoded string? does it encode nicely in a nested way this way? (does the user have to do a separate decode on this value after decoding the parent?)

c-p-i-o replied:
Yes, ncclDumpMap is a map of a map
```
std::unordered_map<
      std::string /* ncclUniqueID */,
      std::unordered_map<std::string, std::string> 
 ```
This becomes JSON rather nicely.

wconstab commented on torch/csrc/distributed/c10d/NCCLUtils.hpp:
```diff
@@ -805,6 +813,77 @@ struct NCCLTraceBuffer {
+    auto result = dump_entries();
+
+    std::list<json> entries;
+    for (auto i : c10::irange(result.size())) {
```
its a bit error-prone to have 2 methods for parsing 'dump_entries()' output and converting their fields one by one.

i'm not sure what to do instead though.

c-p-i-o replied:
I'll see if I can templatize this in a follow-on review.

c-p-i-o committed:
Update on "Flight recoder data as JSON"


Summary:
Provide a new API to retrieve flight recorder data as JSON.
The one minor difference between flight recorder as Pickle v/s JSON is
that the JSON API does not retrieve stack traces at the moment.
This ends up being far too much data.

Test Plan:
unit test

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

c-p-i-o commented:
@c-p-i-o has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59536460).

c-p-i-o commented:
@pytorchbot merge

xwang233 commented:
It seems like this PR added a header `#include <nlohmann/json.hpp>` but didn't properly set the compiler include path for C++ extensions. Our pytorch extension build fails with 
```
#40 42.58 In file included from /opt/pytorch/pytorch/torch/include/torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:22,
#40 42.58                  from xxx.cpp:19:
#40 42.58 /opt/pytorch/pytorch/torch/include/torch/csrc/distributed/c10d/NCCLUtils.hpp:8:10: fatal error: nlohmann/json.hpp: No such file or directory
#40 42.58     8 | #include <nlohmann/json.hpp>
#40 42.58       |          ^~~~~~~~~~~~~~~~~~~
#40 42.58 compilation terminated.
```

I think this is not the first time that pytorch distributed changes break C++ extension builds. Shall we add some CI checks to detect this?

cc @malfet @atalman 

cc @wujingyue @jjsjann123
